src/ImageSharp/Formats/Webp/Lossy/Vp8Encoding.cs

```diff
@@ -376,49 +404,242 @@ private static void InverseTransformHorizontalPass(Vector128<long> t0, Vector128
+
+                // Load ref.
+                var ref0 = Vector128.Create(Unsafe.As<byte, long>(ref referenceRef), 0);
+                var ref1 = Vector128.Create(Unsafe.As<byte, long>(ref Unsafe.Add(ref referenceRef, WebpConstants.Bps)), 0);
+                var ref2 = Vector128.Create(Unsafe.As<byte, long>(ref Unsafe.Add(ref referenceRef, WebpConstants.Bps * 2)), 0);
+                var ref3 = Vector128.Create(Unsafe.As<byte, long>(ref Unsafe.Add(ref referenceRef, WebpConstants.Bps * 3)), 0);
+
+                // 00 01 02 03 *
+                // 10 11 12 13 *
+                // 20 21 22 23 *
+                // 30 31 32 33 *
+                // Shuffle.
+                Vector128<short> srcLow0 = Sse2.UnpackLow(src0.AsInt16(), src1.AsInt16());
+                Vector128<short> srcLow1 = Sse2.UnpackLow(src2.AsInt16(), src3.AsInt16());
+                Vector128<short> refLow0 = Sse2.UnpackLow(ref0.AsInt16(), ref1.AsInt16());
+                Vector128<short> refLow1 = Sse2.UnpackLow(ref2.AsInt16(), ref3.AsInt16());
+
+                // 00 01 10 11 02 03 12 13 * * ...
+                // 20 21 30 31 22 22 32 33 * * ...
+
+                // Convert both to 16 bit.
+                Vector128<byte> src0_16b = Sse2.UnpackLow(srcLow0.AsByte(), Vector128<byte>.Zero);
+                Vector128<byte> src1_16b = Sse2.UnpackLow(srcLow1.AsByte(), Vector128<byte>.Zero);
+                Vector128<byte> ref0_16b = Sse2.UnpackLow(refLow0.AsByte(), Vector128<byte>.Zero);
+                Vector128<byte> ref1_16b = Sse2.UnpackLow(refLow1.AsByte(), Vector128<byte>.Zero);
+
+                // Compute the difference.
+                Vector128<short> row01 = Sse2.Subtract(src0_16b.AsInt16(), ref0_16b.AsInt16());
+                Vector128<short> row23 = Sse2.Subtract(src1_16b.AsInt16(), ref1_16b.AsInt16());
+
+                // First pass.
+                FTransformPass1SSE2(row01, row23, out Vector128<int> v01, out Vector128<int> v32);
+
+                // Second pass.
+                FTransformPass2SSE2(v01, v32, output);
+            }
+            else
+#endif
             {
-                int a0 = tmp[0 + i] + tmp[12 + i];  // 15b
-                int a1 = tmp[4 + i] + tmp[8 + i];
-                int a2 = tmp[4 + i] - tmp[8 + i];
-                int a3 = tmp[0 + i] - tmp[12 + i];
-                output[0 + i] = (short)((a0 + a1 + 7) >> 4);            // 12b
-                output[4 + i] = (short)((((a2 * 2217) + (a3 * 5352) + 12000) >> 16) + (a3 != 0 ? 1 : 0));
-                output[8 + i] = (short)((a0 - a1 + 7) >> 4);
-                output[12 + i] = (short)(((a3 * 2217) - (a2 * 5352) + 51000) >> 16);
+                int i;
+                Span<int> tmp = scratch.Slice(0, 16);
+
+                int srcIdx = 0;
+                int refIdx = 0;
+                for (i = 0; i < 4; i++)
+                {
+                    int d0 = src[srcIdx] - reference[refIdx]; // 9bit dynamic range ([-255,255])
```

Comment:
If you declare these in reverse you should be able to avoid most of the bounds checks.